File: cpp/src/authdog_client.cpp
#include "authdog/authdog_client.h"
#include <cpr/cpr.h>
#include <nlohmann/json.hpp>
#include <iostream>
#include <sstream>

namespace authdog {

class AuthdogClient::Impl {
public:
    ClientConfig config;
    
    Impl(const ClientConfig& cfg) : config(cfg) {
        // Remove trailing slash from base URL
        if (!config.baseUrl.empty() && config.baseUrl.back() == '/') {
            config.baseUrl.pop_back();
        }
    }
    
    UserInfoResponse getUserInfo(const std::string& accessToken) {
        std::string url = config.baseUrl + "/v1/userinfo";
        
        cpr::Header headers = {
            {"Content-Type", "application/json"},
            {"User-Agent", "authdog-cpp-sdk/0.1.0"},
            {"Authorization", "Bearer " + accessToken}
        };
        
        // Add API key if provided
        if (config.apiKey) {
            headers["Authorization"] = "Bearer " + *config.apiKey;
        }
        
        auto response = cpr::Get(
            cpr::Url{url},
            headers,
            cpr::Timeout{config.timeoutSeconds * 1000}
        );
        
        if (response.status_code == 0) {
            throw ApiException("Request failed: " + response.error.message);
        }
        
        if (response.status_code == 401) {
            throw AuthenticationException("Unauthorized - invalid or expired token");
        }
        
        if (response.status_code == 500) {
            try {
                auto errorData = nlohmann::json::parse(response.text);
                if (errorData.contains("error")) {
                    std::string errorMessage = errorData["error"];
                    if (errorMessage == "GraphQL query failed") {
                        throw ApiException("GraphQL query failed");
                    } else if (errorMessage == "Failed to fetch user info") {
                        throw ApiException("Failed to fetch user info");
                    }
                }
            } catch (const nlohmann::json::exception&) {
                // Ignore JSON parsing errors for error responses
            }
        }
        
        if (!response.status_code == 200) {
            throw ApiException("HTTP error " + std::to_string(response.status_code) + ": " + response.text);
        }
        
        try {
            auto jsonData = nlohmann::json::parse(response.text);
            return jsonData.get<UserInfoResponse>();
        } catch (const nlohmann::json::exception& e) {
            throw ApiException("Failed to parse response: " + std::string(e.what()));
        }
    }
};

AuthdogClient::AuthdogClient(const ClientConfig& config) 
    : pImpl(std::make_unique<Impl>(config)) {
}

AuthdogClient::AuthdogClient(const std::string& baseUrl) 
    : AuthdogClient(ClientConfig{baseUrl}) {
}

AuthdogClient::~AuthdogClient() = default;

AuthdogClient::AuthdogClient(AuthdogClient&&) noexcept = default;
AuthdogClient& AuthdogClient::operator=(AuthdogClient&&) noexcept = default;

UserInfoResponse AuthdogClient::getUserInfo(const std::string& accessToken) {
    return pImpl->getUserInfo(accessToken);
}

void AuthdogClient::close() {
    // CPR doesn't require explicit cleanup
}

} // namespace authdog

File: php/src/AuthdogClient.php
<?php

namespace Authdog;

use Authdog\Exceptions\AuthenticationException;
use Authdog\Exceptions\ApiException;
use Authdog\Types\UserInfoResponse;
use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;
use GuzzleHttp\Exception\RequestException;

/**
 * Main client for interacting with Authdog API
 */
class AuthdogClient
{
    private Client $httpClient;
    private string $baseUrl;
    private ?string $apiKey;

    /**
     * Initialize the Authdog client
     *
     * @param string $baseUrl The base URL of the Authdog API
     * @param string|null $apiKey Optional API key for authentication
     * @param array $options Additional Guzzle HTTP client options
     */
    public function __construct(string $baseUrl, ?string $apiKey = null, array $options = [])
    {
        $this->baseUrl = rtrim($baseUrl, '/');
        $this->apiKey = $apiKey;

        $defaultOptions = [
            'base_uri' => $this->baseUrl,
            'timeout' => 10,
            'headers' => [
                'Content-Type' => 'application/json',
                'User-Agent' => 'authdog-php-sdk/0.1.0',
            ],
        ];

        if ($this->apiKey) {
            $defaultOptions['headers']['Authorization'] = 'Bearer ' . $this->apiKey;
        }

        $this->httpClient = new Client(array_merge($defaultOptions, $options));
    }

    /**
     * Get user information using an access token
     *
     * @param string $accessToken The access token for authentication
     * @return UserInfoResponse User information
     * @throws AuthenticationException When authentication fails
     * @throws ApiException When API request fails
     */
    public function getUserInfo(string $accessToken): UserInfoResponse
    {
        try {
            $response = $this->httpClient->get('/v1/userinfo', [
                'headers' => [
                    'Authorization' => 'Bearer ' . $accessToken,
                ],
            ]);

            $data = json_decode($response->getBody()->getContents(), true);
            return new UserInfoResponse($data);

        } catch (RequestException $e) {
            $statusCode = $e->getResponse() ? $e->getResponse()->getStatusCode() : 0;
            $responseBody = $e->getResponse() ? $e->getResponse()->getBody()->getContents() : '';

            if ($statusCode === 401) {
                throw new AuthenticationException('Unauthorized - invalid or expired token');
            }

            if ($statusCode === 500) {
                $errorData = json_decode($responseBody, true);
                if (isset($errorData['error'])) {
                    if ($errorData['error'] === 'GraphQL query failed') {
                        throw new ApiException('GraphQL query failed');
                    } elseif ($errorData['error'] === 'Failed to fetch user info') {
                        throw new ApiException('Failed to fetch user info');
                    }
                }
            }

            throw new ApiException("HTTP error {$statusCode}: {$responseBody}");
        } catch (GuzzleException $e) {
            throw new ApiException('Request failed: ' . $e->getMessage());
        }
    }

    /**
     * Close the HTTP client (for cleanup)
     */
    public function close(): void
    {
        // Guzzle HTTP client doesn't require explicit cleanup
    }
}
